File: Codes/Garnetwzy/234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* first = head;
        ListNode* second = head;
        while(second && second->next) {
            first = first->next;
            second = second->next->next;
        }
        ListNode *head2 = reverse(first);
        while(head2) {
            if(head->val != head2->val)
                return false;
            head = head->next;
            head2 = head2->next;
        } 
        return true;
    }
    
    ListNode* reverse(ListNode *head) {
        ListNode *pre = NULL;
        while(head) {
            ListNode*tmp = head->next;
            head->next = pre;
            pre = head;
            head = tmp;
        }
        return pre;
    }
    
};

File: codes/HellDream/234.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head, fast = head;
        while(fast!=null &&fast.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        if(fast!=null){
            slow.next = reverseList(slow.next);
            slow = slow.next;
        }else{
            slow = reverseList(slow);
        }
        fast = head;
        while(slow!=null){
            if(fast.val!=slow.val) return false;
            slow = slow.next;
            fast = fast.next;
        }
        
        return true;
    }
    
    public ListNode reverseList(ListNode head){
        ListNode prev = null;
        ListNode curr = head;
        
        while(curr!=null){
            ListNode tmp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = tmp;
        }
        return prev;
    }
}